### PR TITLE
[KOA-6044] Bump storybook to v7

### PR DESCRIPTION
**Pull request by anambl**
Remember to include the following changes:

- [ ] `README.md` (If you have created a new component)
- [ ] Component `README.md`
- [ ] Tests
- [ ] Storybook examples created/updated
- [ ] For breaking changes or deprecating components/properties, migration guides added to the description of the PR. If the guide has large changes, consider creating a new Markdown page inside the component's docs folder and link it here

### CODE REVIEW

File: scripts/transpilation/copy-css.js [hunk 1]
```diff
@@ -30,8 +30,7 @@ const cssFiles = execSync('find packages -name "*.css" | grep -v node_modules')
 cssFiles.map((cssFile) => {
   let component;
   const paths = cssFile.split('packages/');
-  const regEx = new RegExp('/');
-  const componentPath = paths[1].split(regEx);
+  const componentPath = paths[1].split('/');
```
Comment (anambl) [inline scripts/transpilation/copy-css.js:33]: ESLint was complaining about using the RegExp constructor instead of using a regex directly

File: scripts/transpilation/copy-types.js [hunk 2]
```diff
@@ -32,8 +32,7 @@ const typeFiles = execSync(
 typeFiles.map((typeFile) => {
   let component;
   const paths = typeFile.split('packages/');
-  const regEx = new RegExp('/');
-  const componentPath = paths[1].split(regEx);
+  const componentPath = paths[1].split('/');
```
Comment (anambl) [inline scripts/transpilation/copy-types.js:35]: ESLint was complaining about using the RegExp constructor instead of using a regex directly

File: packages/bpk-component-modal/src/BpkModalV2/BpKModal.module.scss [hunk 3]
```diff
@@ -49,10 +49,6 @@
 .bpk-modal-wrapper.bpk-modal-polyfill {
   .bpk-modal-backdrop {
     position: fixed;
-    top: 0;
```
Comment (anambl) [inline packages/bpk-component-modal/src/BpkModalV2/BpKModal.module.scss:52]: the `inset` prop does the same and stylelint was complaining about this

File: package.json [hunk 4]
```diff
@@ -115,23 +115,24 @@
...
+    "@storybook/react": "^7.1.1",
+    "@storybook/react-webpack5": "^7.1.1",
     "@testing-library/jest-dom": "^5.16.4",
     "@testing-library/react": "^12.1.0",
     "@testing-library/user-event": "^14.2.0",
     "@types/jest": "^29.5.2",
     "@types/jest-axe": "^3.5.3",
     "@types/react": "^17.0.37",
     "@types/react-dom": "^17.0.11",
+    "@types/webpack-env": "^1.18.1",
```
Comment (anambl) [inline package.json:135]: Due to some dependecies of Storybook having been updated with the v7 migration, we need this [here](https://github.com/Skyscanner/backpack/blob/main/packages/bpk-component-icon/all.tsx#L32)

File: examples/bpk-component-datatable/examples.js [hunk 5]
```diff
@@ -81,6 +83,19 @@ const sortFunc = (rowA, rowB) => {
   return deskA > deskB ? 1 : -1;
 };
 
+const LabelComponent = ({ label }) => <div> This is a {label} </div>;
```
Comment (anambl) [inline examples/bpk-component-datatable/examples.js:86]: Rewriting these due to eslint errors

File: .storybook/webpack.config.js [hunk 6]
```diff
@@ -37,6 +37,12 @@ module.exports = ({ config }) => {
   });
   config.resolve.extensions.push('.tsx');
   config.resolve.extensions.push('.ts');
+  /* eslint-disable-next-line no-param-reassign */
+  config.resolve.alias = {
```
Comment (anambl) [inline .storybook/webpack.config.js:41]: need this to ensure we never have multiple versions of react being used in SB

File: package.json [hunk 7]
```diff
@@ -115,23 +115,24 @@
...
     "@skyscanner/bpk-foundations-web": "^16.1.0",
     "@skyscanner/bpk-svgs": "^16.4.0",
-    "@skyscanner/stylelint-config-skyscanner": "^7.0.0",
-    "@storybook/addon-a11y": "^6.5.14",
-    "@storybook/addon-actions": "^6.5.14",
-    "@storybook/addon-viewport": "^6.5.14",
-    "@storybook/builder-webpack5": "^6.5.16",
-    "@storybook/manager-webpack5": "^6.5.16",
-    "@storybook/react": "^6.5.14",
+    "@skyscanner/stylelint-config-skyscanner": "^10.0.0",
```
Comment (anambl) [inline package.json:121]: Bumping `@skyscanner/stylelint-config-skyscanner` was needed as the SB migration caused some nested dependencies upgrades which resulted it incompatible versions of `prettier` and `eslint-plugin-prettier` being used. The latest versions of `@skyscanner/stylelint-config-skyscanner` and `@skyscanner/eslint-config-skyscanner` solves this issue.

File: .storybook/main.ts [hunk 8]
```diff
@@ -16,14 +16,31 @@
...
     '@storybook/addon-a11y',
     '@storybook/addon-actions',
     '@storybook/addon-viewport',
+    '@storybook/addon-docs',
   ],
-  core: {
-    builder: 'webpack5',
+  framework: {
+    name: '@storybook/react-webpack5',
+    options: {},
```
Comment (olliecurtis) [inline .storybook/main.ts:31]: Does this require to be provided even when it's empty?

File: .storybook/main.ts [hunk 9]
```diff
@@ -16,14 +16,31 @@
...
+    autodocs: false,
+    defaultName: 'Documentation',
+  },
+  typescript: {
+    reactDocgenTypescriptOptions: {
+      shouldIncludePropTagMap: true,
+    },
+  },
+  features: {
+    storyStoreV7: false, // TODO: Remove this. Temporarily opt out of on-demand story loading as it blocks Percy from running tests
```
Comment (olliecurtis) [inline .storybook/main.ts:43]: Do we have a ticket number to be able to refer to in the backlog to fix?

File: .storybook/webpack.config.js (hunk 6 above)
Comment (olliecurtis) [reply .storybook/webpack.config.js:41]: Is this a risk given the version of Node we are on should be deduping or does it still happen regardless?

File: .stylelintrc.json [hunk 10]
```diff
@@ -14,6 +14,9 @@
     }],
     "selector-pseudo-class-no-unknown": [true, {
       "ignorePseudoClasses": ["global"]
-    }]
+    }],
+    "function-url-quotes": null,
+    "scss/operator-no-newline-after": null,
+    "selector-not-notation": null
```
Comment (olliecurtis) [inline .stylelintrc.json:20]: Is this a temp thing or do we expect to turn these off now in our stylelint config?

File: examples/bpk-component-datatable/examples.js [hunk 11]
```diff
@@ -16,6 +16,8 @@
  * limitations under the License.
  */
 
+/* eslint-disable react/prop-types */
```
Comment (olliecurtis) [inline examples/bpk-component-datatable/examples.js:19]: I guess this is temporary due to we are not yet migrating our examples to TS and still using flow things here?

File: packages/bpk-component-infinite-scroll/src/DataSource.js [hunk 12]
```diff
@@ -21,13 +21,11 @@
 class DataSource<T = any> {
   listeners: Array<() => mixed>;
 
-  triggerListeners: (...args: Array<any>) => void;
```
Comment (olliecurtis) [inline packages/bpk-component-infinite-scroll/src/DataSource.js:24]: Out of interest why was this removed? Does it cause any issues?

File: packages/bpk-component-infinite-scroll/src/withInfiniteScroll.js [hunk 13]
```diff
@@ -94,12 +94,9 @@ const withInfiniteScroll = <T: ExtendedProps>(
   ComponentToExtend: AbstractComponent<T>,
 ): AbstractComponent<PropsWithDefault & $Diff<T, ExtendedProps>> =>
   class WithInfiniteScroll extends Component<Props, State> {
-    handleIntersection: IntersectionObserverCallback;
```
Comment (olliecurtis) [inline packages/bpk-component-infinite-scroll/src/withInfiniteScroll.js:97]: Same here

File: packages/bpk-component-infinite-scroll/src/withInfiniteScroll.js [hunk 14]
```diff
@@ -94,12 +94,9 @@ const withInfiniteScroll = <T: ExtendedProps>(
...
 
     handleKeyPress: (e: SyntheticKeyboardEvent<HTMLButtonElement>) => void;
 
-    handleSeeMoreClick: (e: SyntheticEvent<HTMLButtonElement>) => void;
```
Comment (olliecurtis) [inline packages/bpk-component-infinite-scroll/src/withInfiniteScroll.js:101]: Same here

File: scripts/transpilation/copy-css.js (hunk 1 above)
Comment (olliecurtis) [reply scripts/transpilation/copy-css.js:33]: Guessing this was also checked post this change and things are still happy?

Comment (anambl) [reply scripts/transpilation/copy-css.js:33]: I haven't tested it actually 🤔 I thought they're exactly the same thing, but I'll check it!

File: .storybook/main.ts (hunk 8 above)
Comment (anambl) [reply .storybook/main.ts:31]: I'm not sure, it was added automatically by the migration script

File: .storybook/main.ts (hunk 9 above)
Comment (anambl) [reply .storybook/main.ts:43]: no, haven't created a ticket. To be honest we're kinda blocked on this while Storybook sorts out this issue and properly supports Percy

File: .stylelintrc.json (hunk 10 above)
Comment (anambl) [reply .stylelintrc.json:20]: It is temporary, but couldn't add the comment in the json file 😞

File: examples/bpk-component-datatable/examples.js (hunk 11 above)
Comment (anambl) [reply examples/bpk-component-datatable/examples.js:19]: Yes, we can remove this when we migrate to TS

File: packages/bpk-component-infinite-scroll/src/DataSource.js (hunk 12 above)
Comment (anambl) [reply packages/bpk-component-infinite-scroll/src/DataSource.js:24]: Yes! Was complaining that it's defined twice. I don't think we need it here as the method is defined and this was just for the type I assume?

File: .storybook/webpack.config.js (hunk 6 above)
Comment (anambl) [reply .storybook/webpack.config.js:41]: To be honest, I don't understand 100% why there would be multiple copies of react. Running `npm ls react` or `npm ls react-dom` does show only one version and one copy, my assumption is that it's some issue with Storybook

**Review by olliecurtis — APPROVED** [review]
